### DISCUSSION TITLE
Creating Custom Rules

### DISCUSSION THREAD

**Discussion by mergebandit** [graphql-hive/graphql-eslint#1154] [General]
Hey there -

I've been tinkering with creating some custom rules - one example would be ensuring any list types are plural.

If I clone this repo and add my custom rules there, everything works fine. However I'd rather define them in my monorepo (which uses `nx`) ... but I'm running into a problem.

Since `graphql-eslint` uses `patch-package` to patch `eslint`s `rule-tester.js`, I can't just create this as a standard nx lib, which would share the same root package.json / node_modules folder, as the rest of my tests won't get assertions anymore 😂 

What is the general guidance on this? Should I be forking `graphql-eslint` and managing my rules in that forked repo? If I want to have a custom plugin that utilizes `graphql-eslint`, does it need to have a standalone package w/ the postinstall executing `patch-package`? What do other people typically do?

Thanks!

**Comment by dimaMachina** [comment]
Did you read comments in this patch? Do you still need a patch for your project? I don’t think so

**Comment by mergebandit** [comment]
Yeah I read the comments. I'm using `eslint` on a dozen packages within my nx monorepo. If I run _regular_ eslint against my custom rule (or if I just grab one of your rules + tests), I get the `"The rule fixed the code. Please add 'output' property."` error and all of my `invalid` tests fail. If I run w/ the patch, I don't. But I don't want to run the patch since I have a dozen other packages that are also using `eslint`.

So the basic question is ... how do you recommend setting up custom rules? Can I do this as a standalone plugin in my monorepo? Should I do it as a standalone package that consumes / utilizes `graphql-eslint/eslint-plugin`? Should I fork your package and add?

**Reply by dimaMachina** [reply]
You don’t need to clone graphql-eslint in your monorepo, and yes, if you want to share your custom rules it must be as your standalone eslint plugin

From graphql-eslint you should use only graphql parser to parse your graphql document into AST that eslint can understand

**Comment by mergebandit** [comment]
I’m also using the [GraphQLRuleTester](https://github.com/B2o5T/graphql-eslint/blob/master/docs/custom-rules.md#testing-your-rules) and without the eslint patch this is the one that fails with “ The rule fixed the code. Please add 'output' property.”

I guess I’ll just try it with its own patched version of eslint 🤷‍♂️

**Reply by dimaMachina** [reply]
So just add output property. I don’t use output since I already use snapshots testing, for this reason I patch source code

**Comment by mergebandit** [comment]
Alright I’ll try that. Thanks!
